src/core/guardedRequests.slice.ts

```diff
@@ -0,0 +1,143 @@
+  redirectTo,
+  turnUrlStringsIntoObjects
+} from "./utils/helpers/url";
+import { App } from "./utils/types/ids";
+import { userIsAnonymous } from "./utils/helpers/user";
+
+export const AUTH_PARAM = "didAuthenticate";
+
+interface Callback<T1, T2 = void> {
+  (args: T1): T2;
+}
+
+const requestCallbacks: {
+  [key: string]: Callback<Record<string, unknown>>;
+} = {
+  addFavorite: ({ id }) => {
+    return addItem("default", id as string);
+  }
+};
+
+type CallbackType = keyof typeof requestCallbacks;
+
+export type RequestItem = {
+  type: CallbackType;
+  args: Record<string, unknown>;
+  app: App;
+  expire?: number;
+};
+const getExpireTimestamp = () => getCurrentUnixTime() + 60;
+export const hasRequestExpired = ({ expire }: RequestItem) => {
+  // If no expiration timestamp is given. We will expire the request immediately.
+  if (!expire) {
+    return true;
+  }
+  return getCurrentUnixTime() > expire;
+};
+
+const initialState: { request: RequestItem | null } = { request: null };
+const guardedRequests = createSlice({
+  name: "guardedRequests",
+  initialState,
+  reducers: {
+    addRequest(state, action: PayloadAction<RequestItem>) {
+      const { payload: request } = action;
+      if (!request?.expire) {
+        request.expire = getExpireTimestamp();
+      }
+      state.request = request;
+    },
+    removeRequest(state) {
+      state.request = null;
+    }
+  }
+});
+export const { addRequest, removeRequest } = guardedRequests.actions;
+
+const getRequestCallback = (
+  type: CallbackType
+): Callback<Record<string, unknown>> => requestCallbacks?.[type];
+
+const requestCallbackExists = (type: CallbackType) =>
+  Boolean(getRequestCallback(type));
+
+const getUrlsFromState = (state: RootState) => {
+  const {
+    url: { data }
+  } = state;
+  return turnUrlStringsIntoObjects(data);
+};
+
+export const guardedRequest = createAsyncThunk(
+  "guardedRequests/performRequest",
+  async (
+    requestItem: RequestItem,
+    { dispatch, fulfillWithValue, getState }
+  ) => {
+    const { type, args } = requestItem;
+    // The callback is unknown and we cannot continue.
+    if (!requestCallbackExists(type)) {
+      return fulfillWithValue({ status: "ignored", message: "Nothing to do" });
+    }
+
+    // User is anonymous and the requests is known.
+    if (userIsAnonymous()) {
+      // So we'll store the request for later execution.
+      dispatch(addRequest(requestItem));
+
+      // Make sure that the request is persisted.
+      persistor.flush().then(() => {
+        // And redirect to external login.
+        const { authUrl } = getUrlsFromState(getState() as RootState);
+        if (authUrl) {
+          const { pathname, search } = appendQueryParametersToUrl(
+            new URL(getCurrentLocation()),
+            {
+              [AUTH_PARAM]: "1"
+            }
+          );
+          const redirectUrl = appendQueryParametersToUrl(authUrl, {
+            "current-path": `${pathname}${search}`
+          });
```

Comment:
Consider if this can be moved to the mounting application in the `authUrl` prop.

It seems as if this will create a hard dependency between whatever is located at `authUrl` (the authenticating app) and the guarded app. The authenticating app has to store and redirect to whatever is provided in the current path query parameter.

How about instead stating that it is the responsibility of the mounting app to ensure that whatever url is provided as `authUrl` will return the user to an appropriate state. In our case Drupal can provide `/oauth/adgangsplatformen/redirect?destination=/search?q=harry` as auth url and this will ensure that the browser ends up at `/search?q=harry` after authentication. When the guarded app can pick up where it left off.

Reply:
It's a part of the openid_connect auth flow to delete the destination, that's why I added the current-path functionality.
I don't consider the current-path principle to be a hard dependency between the app and Drupal. Current path is just an extra feature the Drupal endpoint exposes that makes it possible to redirect to it after ended authentication. Actually it is a replacement of the Drupal destination feature since it was removed by the openid_connect module.